sbi/inference/trainers/npe/npe_b.py

```diff
@@ -23,54 +23,80 @@ def __init__(
+                needs to return a PyTorch `nn.Module` implementing the density
+                estimator. The density estimator needs to provide the methods
+                `.log_prob` and `.sample()`.
+            device: Training device, e.g., "cpu", "cuda" or "cuda:{0, 1, ...}".
+            logging_level: Minimum severity of messages to log. One of the strings
+                INFO, WARNING, DEBUG, ERROR and CRITICAL.
+            summary_writer: A tensorboard `SummaryWriter` to control, among others, log
+                file location (default is `<current working directory>/logs`.)
+            show_progress_bars: Whether to show a progressbar during training.
+        """
 
         kwargs = del_entries(locals(), entries=("self", "__class__"))
         super().__init__(**kwargs)
 
     def _log_prob_proposal_posterior(
-        self, theta: Tensor, x: Tensor, masks: Tensor
+        self,
+        theta: Tensor,
+        x: Tensor,
+        masks: Tensor,
+        proposal: Optional[Any],
     ) -> Tensor:
         """
         Return importance-weighted log probability (Lueckmann, Goncalves et al., 2017).
 
         Args:
             theta: Batch of parameters θ.
             x: Batch of data.
-            masks: Whether to retrain with prior loss (for each prior sample).
+            masks: Mask that is True for prior samples in the batch in order to train
+                them with prior loss.
+            proposal: Proposal distribution.
 
         Returns:
-            Log probability of proposal posterior.
+            Importance-weighted log-probability of the proposal posterior.
         """
 
-        raise NotImplementedError
-
-        batch_size = theta.shape[0]
-
-        # Evaluate posterior.
-        log_prob_posterior = self._posterior.net.log_prob(theta, x)
-        log_prob_posterior = log_prob_posterior.reshape(batch_size)
-        utils.assert_all_finite(log_prob_posterior, "posterior eval")
+        # Evaluate prior
+        utils.assert_all_finite(self._prior.log_prob(theta), "prior eval")
+        prior = torch.exp(self._prior.log_prob(theta))
 
-        # Evaluate prior.
-        log_prob_prior = self._prior.log_prob(theta).reshape(batch_size)
-        utils.assert_all_finite(log_prob_prior, "prior eval.")
+        # Evaluate proposal
+        # (as theta comes from prior and proposal from previous rounds,
+        # the last proposal is actually a mixture of the prior
+        # and of all the previous proposals with coefficients 1/round)
+        prop = 1.0 / (self._round + 1)
+        proposal = torch.zeros(theta.size(0), device=theta.device)
 
-        # Evaluate proposal.
-        log_prob_proposal = self._model_bank[-1].net.log_prob(theta, x)
-        utils.assert_all_finite(log_prob_proposal, "proposal posterior eval")
+        for density in self._proposal_roundwise:
+            utils.assert_all_finite(density.log_prob(theta), "proposal eval")
```

Comment:
see comment above regarding error message. Here, I agree that if density.log_prob is -inf then this should be an error.